lib/plugins/aws/deployFunction/tests/index.js

```diff
@@ -0,0 +1,245 @@
+'use strict';
+
+const expect = require('chai').expect;
+const sinon = require('sinon');
+const path = require('path');
+const fs = require('fs');
+const os = require('os');
+const Zip = require('node-zip');
+const AwsDeployFunction = require('../');
+const Serverless = require('../../../../Serverless');
+const BbPromise = require('bluebird');
+
+describe('AwsDeployFunction', () => {
+  let serverless;
+  let awsDeployFunction;
+
+  beforeEach(() => {
+    serverless = new Serverless();
+    serverless.servicePath = true;
+    serverless.service.environment = {
+      vars: {},
+      stages: {
+        dev: {
+          vars: {},
+          regions: {
+            'us-east-1': {
+              vars: {},
+            },
+          },
+        },
+      },
+    };
+    serverless.service.functions = {
+      first: {
+        handler: true,
+      },
+    };
+    const options = {
+      stage: 'dev',
+      region: 'us-east-1',
+      function: 'first',
+    };
+    serverless.init();
+    awsDeployFunction = new AwsDeployFunction(serverless, options);
+  });
+
+  describe('#constructor()', () => {
+    it('should have hooks', () => expect(awsDeployFunction.hooks).to.be.not.empty);
+
+    it('should set the provider variable to "aws"', () => expect(awsDeployFunction.provider)
+      .to.equal('aws'));
+
+    it('should run promise chain in order', () => {
+      const checkIfFunctionExistsInServiceStub = sinon
+        .stub(awsDeployFunction, 'checkIfFunctionExistsInService').returns(BbPromise.resolve());
+      const checkIfFunctionIsDeployedStub = sinon
+        .stub(awsDeployFunction, 'checkIfFunctionIsDeployed').returns(BbPromise.resolve());
+      const zipFunctionStub = sinon
+        .stub(awsDeployFunction, 'zipFunction').returns(BbPromise.resolve());
+      const deployFunctionStub = sinon
+        .stub(awsDeployFunction, 'deployFunction').returns(BbPromise.resolve());
+
+      return awsDeployFunction.hooks['deploy:function:deploy']().then(() => {
+        expect(checkIfFunctionExistsInServiceStub.calledOnce).to.be.equal(true);
+        expect(checkIfFunctionIsDeployedStub.calledAfter(checkIfFunctionExistsInServiceStub))
+          .to.be.equal(true);
+        expect(zipFunctionStub.calledAfter(checkIfFunctionIsDeployedStub))
+          .to.be.equal(true);
+        expect(deployFunctionStub.calledAfter(zipFunctionStub))
+          .to.be.equal(true);
+
+        awsDeployFunction.checkIfFunctionExistsInService.restore();
+        awsDeployFunction.checkIfFunctionIsDeployed.restore();
+        awsDeployFunction.zipFunction.restore();
+        awsDeployFunction.deployFunction.restore();
+      });
+    });
+  });
+
+  describe('#checkIfFunctionExistsInService()', () => {
```

Comment:
unnecessary `describe` block imo ... if we move the validation somewhere else we won't need it.